src/cpp/shared/game/game.cpp

```diff
@@ -268,130 +267,124 @@ void game::ProcessOnMapHeroes() {
   hero *randomHero;
   char mightBeHeroAlreadyExists[56];
   int ppMapExtraHeroIdx;
-  heroMapExtra *mapExtraHero;
+  HeroExtra *mapExtraHero;
   char isJail;
   int i;
 
-  result = memset(mightBeHeroAlreadyExists, 0, 54u);// Only 54 out of 56 bytes set to zero, for whatever reason.
-  for (i = 0; i < 3; ++i) {
-    for (coordYForRandomHero = 0; ; ++coordYForRandomHero) {
-      result = (void *)MAP_HEIGHT;
-      if (coordYForRandomHero >= MAP_HEIGHT)
-        break;
-      for (coordXForRandomHero = 0; coordXForRandomHero < MAP_WIDTH; ++coordXForRandomHero) {
-        loc = &this->map.tiles[(coordYForRandomHero * this->map.width) + coordXForRandomHero];
-        if ((loc->objType & 0x7F) == 55 || loc->objType == 251) {
-          isJail = (loc->objType & 0x7F) == LOCATION_JAIL;
-          ppMapExtraHeroIdx = loc->extraInfo;
-          mapExtraHero = (heroMapExtra *)ppMapExtra[ppMapExtraHeroIdx];
-
-          if (!i) {                           // 1st iteration
-            if (!mapExtraHero->field_11 || mapExtraHero->heroID >= 54 || mightBeHeroAlreadyExists[mapExtraHero->heroID]) {
-              mapExtraHero->couldBeHasFaction = 0;
-            } else {
-              mightBeHeroAlreadyExists[mapExtraHero->heroID] = 1;
-              mapExtraHero->couldBeHasFaction = 1;
-            }
-            if (isJail) {
-              mapExtraHero->owner = -1;
-            } else {
-              mapExtraHero->owner = loc->objectIndex / 7; // Uh, redundant line?
-              mapExtraHero->owner = gcColorToPlayerPos[mapExtraHero->owner];
-            }
+  memset(mightBeHeroAlreadyExists, 0, 54u);// Only 54 out of 56 bytes set to zero, for whatever reason.
+  for (coordYForRandomHero = 0; coordYForRandomHero < MAP_HEIGHT; ++coordYForRandomHero) {
+    for (coordXForRandomHero = 0; coordXForRandomHero < MAP_WIDTH; ++coordXForRandomHero) {
+      loc = &this->map.tiles[(coordYForRandomHero * this->map.width) + coordXForRandomHero];
+      if ((loc->objType & 0x7F) == 55 || loc->objType == 251) {
```

Comment:
Have we talked about how this field works, where the bottom 7 bits represent the location type, and the top bit represents if this is a location with an event? It would be nice to replace this one field with two fields (of size 7 bits and 1 bit), except they sometimes write this field as if it's one field. That may just be a compiler optimization though.

We should probably replace the "& 0x7F" with a GET_LOCATION_TYPE() macro. You can then replace the 55 and 251 with constants from the ADVENTURE_MAP_LOCATION (or whatever it's called) enum.

Reply:
I recall discussing this somewhat. Also, I'm not entirely sure I understand that macro. I think I understand what it might do, but not how it would work.

Reply:
#define GET_LOCATION_TYPE(n) ((n) & 0x7F)

Reply:
Oh, ok. So, I would just replace "n" with the actual ADVENTURE_MAP_LOCATION for that context?

Reply:
You'd invoke this as GET_LOCATION_TYPE(loc->objType) .

Reply:
Or maybe you'd define it so you can just do GET_LOCATION_TYPE(loc).

Or define loc as a class, so you can say loc->getLocationType(); 

Point is, don't spread the knowledge that the location type is stored in the bottom 7 bits all over the codebase. That will make it even more impossible to change.

Reply:
I like the idea of the getLocationType() function. "loc" is of type mapCell*. This would just be adding this function to the class definition, I think.